Feature/ericlow/zns info panel

ericlowturing opened the pull request:
<!-- NOTE: Please use the following template - it makes the reviewer's lives much easier! -->

[Associated Notion Card](https://www.notion.so/dcf93eee667d4edeac4b2a927d211551?v=4e41756a49994938b31a761c9432eb4f&p=34957a7d267e43f68a137fd5818e7164)

-----

## 1. Pull request checklist

<!-- Please make sure to do the following - your PR may not be accepted if any of these aren't completed: -->
- [ ] Notion card has been moved to the Code Review column
- [ ] Notion card has a link to this PR
- [ ] A reviewer has been assigned to the Notion card


## 2. PR type
<!-- 
  Please try to limit your pull request to one type, submit multiple pull requests if needed. 
  One of:
    - Bugfix
    - Feature
    - Code style update (formatting, renaming)
    - Refactoring (no functional changes, no api changes)
    - Build related changes
    - Documentation content changes
    - Other (please describe):
--> 


## 3. What is the old behaviour?
<!-- Please describe the old behaviour that you are modifying. -->


## 4. What is the new behaviour?
<!-- Please describe the behaviour or changes that are being added by this PR. -->


## 5. Other information
<!-- Optional: any other information that is important to this PR such as a Loom or screenshots describing behaviour outlined in Step 4. -->

ericlowturing commented on src/containers/PageContainer/elements/Header/elements/InfoButton/elements/InfoPanel/InfoPanel.constants.tsx:
```diff
@@ -0,0 +1,97 @@
+export const ZNS_INFO_OPTIONS = {
+	[ZNS_INFO_KEYS.DISCORD]: {
+		label: 'Discord',
+		link: 'https://discord.zero.com',
```
@stacy-hunt Can you help me to get discord link here? Thanks

ericlowturing commented on src/containers/PageContainer/index.tsx:
```diff
@@ -295,129 +225,23 @@ const PageContainer: FC = ({ children }) => {
-						)}
-					</>
-				</TitleBar>
+
+				<Header
+					pageWidth={pageWidth}
+					znsDomain={znsDomain}
+					account={account}
+					openModal={openModal}
+				/>
```
@stacy-hunt @colbr @domw30 @padala @akshatmittal This is just one element used for header component. We used so many codes here which is hard to read

domw30 replied:
looks good, nice!

ericlowturing commented on src/constants/localStorage.ts:
```diff
@@ -0,0 +1,4 @@
+export enum LOCAL_STORAGE_KEYS {
+	// ----- CHOOSEN_WALLET ----- //
+	CHOOSEN_WALLET = '/chosenWallet',
+}
```
@colbr @padala @domw30 @akshatmittal We will use global constants for local storage keys instead of hard coding

ericlowturing commented on src/constants/routes.ts:
```diff
@@ -0,0 +1,9 @@
+export enum ROUTES {
+	// ----- Market ----- //
+	MARKET = '/market',
+
+	// ----- STAKING ----- //
+	STAKING = '/staking',
+	STAKING_POOLS = '/pools',
+	STAKING_DEPOSITS = '/deposits',
+}
```
@colbr @padala @domw30 @akshatmittal We will use global constants for routes instead of hard coding in each component

ericlowturing commented on src/constants/urls.ts:
```diff
@@ -0,0 +1,4 @@
+export enum URLS {
+	// ----- WILDERWORLD ----- //
+	WILDERWORLD = 'https://www.wilderworld.com/',
+}
```
@colbr @padala @domw30 @akshatmittal We will start using global constants for external URLs instead of hard coding in each component

domw30 reviewed (commented):
This is nice and clean 👍 thanks for tagging Eric

domw30 commented on src/containers/PageContainer/elements/Header/elements/ConnectWalletButton/_connect-wallet-button.scss:
```diff
@@ -0,0 +1,11 @@
+	}
+
+	&__title {
+		margin: 0 1rem;
```
rem or pixels?

ericlowturing replied:
I definitely resolve this!

domw30 commented on src/containers/PageContainer/elements/Header/elements/ConnectWalletButton/hooks/useConnectWalletButtonData.ts:
```diff
@@ -0,0 +1,49 @@
+		const isConnecting = !isConnected && Boolean(wallet);
+		const connectTitle = isConnecting
+			? `Trying to connect ${wallet}`
+			: `Connect ${props.isDesktop && 'Wallet'}`;
```
Did you want to remove all hardcoded values and add as constants?

ericlowturing replied:
Hmm, this title was defined in `JSX` and I moved it into custom hooks. Not sure it is constant, it is detecting desktop or mobile, and set the title.

domw30 commented on src/containers/PageContainer/elements/Header/elements/HistoryButtons/HistoryButtons.tsx:
```diff
@@ -0,0 +1,52 @@
+				onClick={handlers.handleOnForwardClick}
+				className="history-buttons__button"
+				disabled={!formattedData.canGoForward}
+				alt="Click to go forward"
```
We could extract the alt text here

ericlowturing replied:
Just copied this text from `Chrome browser history buttons`

domw30 commented on src/containers/PageContainer/elements/Header/elements/InfoButton/elements/InfoPanel/_info-panel.scss:
```diff
@@ -0,0 +1,97 @@
+.info-panel {
+	&__container {
+		position: absolute;
```
pixels and rem

ericlowturing replied:
I think we need to use only one of them pixel or rem. I always used rem because it should be responsive in all of screens.

domw30 commented on src/containers/PageContainer/elements/Header/elements/StatusButtons/StatusButtons.tsx:
```diff
@@ -0,0 +1,61 @@
+			{showTransferringButton && (
+				<TooltipLegacy content={<TransferPreview />}>
+					<NumberButton
+						rotating={statusCounts.transferring > 0}
```
same here

domw30 commented on src/containers/PageContainer/elements/Header/elements/InfoButton/elements/InfoPanel/InfoPanel.tsx:
```diff
@@ -0,0 +1,99 @@
+								key={`wilder-options-${option.link}`}
+								onClick={onClose}
+							>
+								<img src={option.icon} alt={option.label} />
```
I also think we should add title attribute to img tags so a user can hover over and see what the icon is for

ericlowturing replied:
Agreed. Design had a tooltip at first, and I believe it is good to have tooltip in UI/UX

domw30 commented:
@ericlowturing I realised most of my comments above are very similar to what you previously posted earlier 😬 please ignore any that you addressed earlier !

ericlowturing commented on src/components/Buttons/NumberButton/NumberButton.tsx:
```diff
@@ -6,7 +6,7 @@ import styles from './NumberButton.module.scss';
 type NumberButtonProps = {
 	number: number;
-	onClick: () => void;
+	onClick?: () => void;
```
Yes, it is needed. For disabled button (i.e: no minting/skating button) shouldn't have `onClick`

ericlowturing commented on src/containers/PageContainer/elements/Header/_header.scss:
```diff
@@ -0,0 +1,102 @@
+
+		&-title {
+			margin-left: 16px;
+			font-size: 1rem;
```
yep, agreed. I copied styles from old `TitleBar.module.scss` due to timing. But I recommend to use `rem/em` for font size

ericlowturing commented on src/containers/PageContainer/elements/Header/_header.scss:
```diff
@@ -0,0 +1,102 @@
+			flex: 1;
+			margin-left: 16px;
+			background: none;
+			color: white;
```
yep, agreed. I copied styles from old `TitleBar.module.scss` due to timing.

ericlowturing commented on src/containers/PageContainer/elements/Header/_header.scss:
```diff
@@ -0,0 +1,102 @@
+			font-weight: bold;
+
+			&::placeholder {
+				color: #e5baff70;
```
yep, agreed. I copied styles from old `TitleBar.module.scss` due to timing.

ericlowturing commented on src/containers/PageContainer/elements/Header/_header.scss:
```diff
@@ -0,0 +1,102 @@
+			box-shadow: none;
+		}
+
+		@media only screen and (max-width: 1000px) {
```
Nice 👍

ericlowturing commented on src/containers/PageContainer/elements/Header/elements/ConnectWalletButton/ConnectWalletButton.tsx:
```diff
@@ -0,0 +1,58 @@
+	onConnectWallet: () => void;
+};
+
+export const ConnectWalletButton: React.FC<ConnectWalletButtonProps> = ({
```
Since we are using typescript, we should specify it is React component. Your code looks like more `javascript` and looks like just `hooks` :)

ericlowturing commented on src/containers/PageContainer/elements/Header/elements/InfoButton/_info-button.scss:
```diff
@@ -0,0 +1,28 @@
+	&:hover {
+		background: linear-gradient(
+			90deg,
+			rgba(82, 87, 255, 0.25) 0%,
```
Agreed, but I have no idea where is our global colors in scss. I just copied from figma for now

ericlowturing commented on src/containers/PageContainer/elements/Header/elements/InfoButton/elements/InfoPanel/InfoPanel.constants.tsx:
```diff
@@ -0,0 +1,93 @@
+	ZERO = 'zero',
+	ZINE = 'zine',
+}
+
```
It is just `enum`?

ericlowturing commented on src/containers/PageContainer/elements/Header/elements/StatusButtons/StatusButtons.tsx:
```diff
@@ -0,0 +1,61 @@
+			{/* Status / Long Running Operation Button */}
+			{showStatusButton && (
+				<TooltipLegacy content={<MintPreview onOpenProfile={onOpenProfile} />}>
+					<NumberButton rotating={statusCount > 0} number={statusCount} />
```
Yep, agreed. But I decided to use this simple condition checking here because it is not hard to read.

ericlowturing replied:
Super simple condition checking

colbr commented:
I just had a quick skim of this one as I'm quite tied up, but it seems to be working well in the dApp. One small issue is [the pop-up doesn't have a background on Firefox](https://ibb.co/0VthfFq). This is actually a problem with a bunch of elements throughout the dApp (has been for a while), but if you could fix it just for the info panel for now that would be awesome.

I also noticed you aren't using CSS modules - is there any reason for this? We have been using modules across the code base. If you don't think we should use them I am happy to hear why.

A bigger problem - I have noticed your custom hooks are split into separate `data`, `handlers`, and `lifecycle` hooks. I appreciate the value encapsulated logic, however, I have a couple of problems with how far it has been taken here:
1. Hooks seems to be used arbitrarily. I will use `InfoPanel.tsx` as an example. The `onConnectWallet` (function) prop is sent to `useInfoPanelHandlers`, which returns an object containing the same function it was passed - this function returned from the hook is then fed into `<ConnectWalletButton>`. I think this is overly-complex, as the `onConnectWallet` could have been passed straight to `<ConnectWalletButton>`. I think a custom hook should only be used when (1) a component is getting too complex, and/or (2) when some logic will be re-used somewhere else in the codebase. 
2. Custom hooks are already a means to encapsulate grouped functionality, so I am not sure what is gained by splitting a hook into more layers - in this case `data`, `handlers`, and `lifecycle`. I think the code would be much easier to read, understand, and debug if there was one custom hook per group of functionality, i.e. `useConnectWallet`.

To summarise the above paragraph - I think we should use hooks when (1) a component is getting complex, and (2) when some logic needs to be re-used somewhere else. I don't think `InfoPanel` needs a hook, and of the components which do need hooks, I don't think any of them are complex enough to split into separate layers.  Code readability is very important for productivity, and I think this should be taken into account when considering how to approach a problem.

Will tag @padala, and @domw30 to get their thoughts on the above! For reference, [this is my idea of how hooks should be used](https://github.com/zer0-os/zNS-dapp/tree/feat/cancel-bid/src/containers/flows/CancelBid) 😊

ericlowturing commented:
> I just had a quick skim of this one as I'm quite tied up, but it seems to be working well in the dApp. One small issue is [the pop-up doesn't have a background on Firefox](https://ibb.co/0VthfFq). This is actually a problem with a bunch of elements throughout the dApp (has been for a while), but if you could fix it just for the info panel for now that would be awesome.
> 
> I also noticed you aren't using CSS modules - is there any reason for this? We have been using modules across the code base. If you don't think we should use them I am happy to hear why.
> 
> A bigger problem - I have noticed your custom hooks are split into separate `data`, `handlers`, and `lifecycle` hooks. I appreciate the value encapsulated logic, however, I have a couple of problems with how far it has been taken here:
> 
> 1. Hooks seems to be used arbitrarily. I will use `InfoPanel.tsx` as an example. The `onConnectWallet` (function) prop is sent to `useInfoPanelHandlers`, which returns an object containing the same function it was passed - this function returned from the hook is then fed into `<ConnectWalletButton>`. I think this is overly-complex, as the `onConnectWallet` could have been passed straight to `<ConnectWalletButton>`. I think a custom hook should only be used when (1) a component is getting too complex, and/or (2) when some logic will be re-used somewhere else in the codebase.
> 2. Custom hooks are already a means to encapsulate grouped functionality, so I am not sure what is gained by splitting a hook into more layers - in this case `data`, `handlers`, and `lifecycle`. I think the code would be much easier to read, understand, and debug if there was one custom hook per group of functionality, i.e. `useConnectWallet`.
> 
> To summarise the above paragraph - I think we should use hooks when (1) a component is getting complex, and (2) when some logic needs to be re-used somewhere else. I don't think `InfoPanel` is complex enough to need a hook, and of the components which do need hooks, I don't think any of the hooks are complex enough to justify splitting into separate layers. Code readability is very important for productivity, and I think this should be taken into account when considering how to approach a problem.
> 
> Will tag @padala, and @domw30 to get their thoughts on the above! For reference, [this is my idea of how hooks should be used](https://github.com/zer0-os/zNS-dapp/tree/feat/cancel-bid/src/containers/flows/CancelBid) 😊
> 
> p.s. this turned into a bit of a paragraph, apologies!

Hi @colbr Thanks for your reviewing. I agreed it is super simple component - `InfoPanel.tsx`. Because it is just sending back `props` function. I just read it again. but it was 2 handlers function before I push last commit with css changes. And there was `useInfoPanelData.tsx` ass well. I totally agreed it is complex the readbility of `infoPanel.tsx` right now. But other components, I think it is not. For the component, we will just have  `component - level ``hooks` and `jsx` and we will handle `data, callback functions, lifecycles` in the `component level custom hooks`. And the `logical functions`  will be used in `component level helpers` if the logic is used in the component only, or will be used in `global helpers` which will be used in other components. And I also agreed to use `global hooks` like which is accessing `redux or provider`.

ericlowturing commented:
> > I just had a quick skim of this one as I'm quite tied up, but it seems to be working well in the dApp. One small issue is [the pop-up doesn't have a background on Firefox](https://ibb.co/0VthfFq). This is actually a problem with a bunch of elements throughout the dApp (has been for a while), but if you could fix it just for the info panel for now that would be awesome.
> > I also noticed you aren't using CSS modules - is there any reason for this? We have been using modules across the code base. If you don't think we should use them I am happy to hear why.
> > A bigger problem - I have noticed your custom hooks are split into separate `data`, `handlers`, and `lifecycle` hooks. I appreciate the value encapsulated logic, however, I have a couple of problems with how far it has been taken here:
> > 
> > 1. Hooks seems to be used arbitrarily. I will use `InfoPanel.tsx` as an example. The `onConnectWallet` (function) prop is sent to `useInfoPanelHandlers`, which returns an object containing the same function it was passed - this function returned from the hook is then fed into `<ConnectWalletButton>`. I think this is overly-complex, as the `onConnectWallet` could have been passed straight to `<ConnectWalletButton>`. I think a custom hook should only be used when (1) a component is getting too complex, and/or (2) when some logic will be re-used somewhere else in the codebase.
> > 2. Custom hooks are already a means to encapsulate grouped functionality, so I am not sure what is gained by splitting a hook into more layers - in this case `data`, `handlers`, and `lifecycle`. I think the code would be much easier to read, understand, and debug if there was one custom hook per group of functionality, i.e. `useConnectWallet`.
> > 
> > To summarise the above paragraph - I think we should use hooks when (1) a component is getting complex, and (2) when some logic needs to be re-used somewhere else. I don't think `InfoPanel` is complex enough to need a hook, and of the components which do need hooks, I don't think any of the hooks are complex enough to justify splitting into separate layers. Code readability is very important for productivity, and I think this should be taken into account when considering how to approach a problem.
> > Will tag @padala, and @domw30 to get their thoughts on the above! For reference, [this is my idea of how hooks should be used](https://github.com/zer0-os/zNS-dapp/tree/feat/cancel-bid/src/containers/flows/CancelBid) 😊
> > p.s. this turned into a bit of a paragraph, apologies!
> 
> Hi @colbr Thanks for your reviewing. I agreed it is super simple component - `InfoPanel.tsx`. Because it is just sending back `props` function. I just read it again. but it was 2 handlers function before I push last commit with css changes. And there was `useInfoPanelData.tsx` ass well. I totally agreed it is complex the readbility of `infoPanel.tsx` right now. But other components, I think it is not. For the component, we will just have `` component - level ``hooks `` and `jsx` and we will handle `data, callback functions, lifecycles` in the `component level custom hooks`. And the `logical functions` will be used in `component level helpers` if the logic is used in the component only, or will be used in `global helpers` which will be used in other components. And I also agreed to use `global hooks` like which is accessing `redux or context`.

And for the CSS modules, I can see many components in our repo are using inline styles and class names in the `JSX` which is bad patter. We need to use only one of them `className` or `inline-style`. Of course, we will choose `className`. And we moved to `.scss` which provides a lot of features like `variables/mixins/siblings`...etc. But when I see our `.scss` files, we are using `var(---variable)` which is css variable stuffs. I think its time to use global variables and mixins, and use/import in our each .scss files.

ericlowturing commented:
> > > I just had a quick skim of this one as I'm quite tied up, but it seems to be working well in the dApp. One small issue is [the pop-up doesn't have a background on Firefox](https://ibb.co/0VthfFq). This is actually a problem with a bunch of elements throughout the dApp (has been for a while), but if you could fix it just for the info panel for now that would be awesome.
> > > I also noticed you aren't using CSS modules - is there any reason for this? We have been using modules across the code base. If you don't think we should use them I am happy to hear why.
> > > A bigger problem - I have noticed your custom hooks are split into separate `data`, `handlers`, and `lifecycle` hooks. I appreciate the value encapsulated logic, however, I have a couple of problems with how far it has been taken here:
> > > 
> > > 1. Hooks seems to be used arbitrarily. I will use `InfoPanel.tsx` as an example. The `onConnectWallet` (function) prop is sent to `useInfoPanelHandlers`, which returns an object containing the same function it was passed - this function returned from the hook is then fed into `<ConnectWalletButton>`. I think this is overly-complex, as the `onConnectWallet` could have been passed straight to `<ConnectWalletButton>`. I think a custom hook should only be used when (1) a component is getting too complex, and/or (2) when some logic will be re-used somewhere else in the codebase.
> > > 2. Custom hooks are already a means to encapsulate grouped functionality, so I am not sure what is gained by splitting a hook into more layers - in this case `data`, `handlers`, and `lifecycle`. I think the code would be much easier to read, understand, and debug if there was one custom hook per group of functionality, i.e. `useConnectWallet`.
> > > 
> > > To summarise the above paragraph - I think we should use hooks when (1) a component is getting complex, and (2) when some logic needs to be re-used somewhere else. I don't think `InfoPanel` is complex enough to need a hook, and of the components which do need hooks, I don't think any of the hooks are complex enough to justify splitting into separate layers. Code readability is very important for productivity, and I think this should be taken into account when considering how to approach a problem.
> > > Will tag @padala, and @domw30 to get their thoughts on the above! For reference, [this is my idea of how hooks should be used](https://github.com/zer0-os/zNS-dapp/tree/feat/cancel-bid/src/containers/flows/CancelBid) 😊
> > > p.s. this turned into a bit of a paragraph, apologies!
> > 
> > 
> > Hi @colbr Thanks for your reviewing. I agreed it is super simple component - `InfoPanel.tsx`. Because it is just sending back `props` function. I just read it again. but it was 2 handlers function before I push last commit with css changes. And there was `useInfoPanelData.tsx` ass well. I totally agreed it is complex the readbility of `infoPanel.tsx` right now. But other components, I think it is not. For the component, we will just have `component - level`hooks `` and `jsx` and we will handle `data, callback functions, lifecycles` in the `component level custom hooks`. And the `logical functions` will be used in `component level helpers` if the logic is used in the component only, or will be used in `global helpers` which will be used in other components. And I also agreed to use `global hooks` like which is accessing `redux or context`.
> 
> And for the CSS modules, I can see many components in our repo are using inline styles and class names in the `JSX` which is bad patter. We need to use only one of them `className` or `inline-style`. Of course, we will choose `className`. And we moved to `.scss` which provides a lot of features like `variables/mixins/siblings`...etc. But when I see our `.scss` files, we are using `var(---variable)` which is css variable stuffs. I think its time to use global variables and mixins, and use/import in our each .scss files.

And thanks for your report the bug on firefox, I just fixed for `BuyTokenRedirect` and `InfoPanel` component.

colbr commented:
> it was 2 handlers function before I push last commit

Ohh ok that makes sense! Sounds good, thanks for fixing.

> or the component, we will just have component-level hooks and jsx and we will handle data, callback functions, lifecycles in the component level custom hooks

Right, I'm just wondering where the benefits are when splitting one custom hook into three smaller hooks - `data`, `lifecycle`, and `handlers`. When you split a hook by `data`, `handlers`, and `lifecycle` the code is **very** encapsulated, but in my opinion it's so encapsulated that it's hard to tell how a component works as a whole.

If your concern is splitting hooks/logic from JSX, we could implement the container-component pattern, where the container handles all of the business logic and the component is purely presentational?

Code readability is very important for productivity, so I'm concerned this pattern is going to move us in the wrong direction by adding friction to the dev workflow. If there are any objective reasons that this pattern is better I think we can look at implementing it for new components. Similarly, if @padala and @domw30 think it is sufficiently readable, we can look into it.

> I can see many components in our repo are using inline styles and class names

Yeah there are a few outdated components in there - those definitely need tidying up. We should be using `className` instead of inline styling.

> our .scss files, we are using var(---variable)

We are, this is because we were using CSS up until recently. `main.scss` should be updated with SCSS variables, but there have been more pressing features to work on.

re: CSS modules - are you happy to use them? Just to be consistent with newer components. Happy to hear why standard SCSS might be better though

ericlowturing commented:
> re: CSS modules - are you happy to use them? Just to be consistent with newer components. Happy to hear why standard SCSS might be better though

Oh, I just read about [CSS modules](https://github.com/css-modules/css-modules) and I am first at CSS Modules. I am wondering if there would be some conflicts or issues with `scss`. But I will use `CSS MODULE` from now. Thanks

padala commented:
> Code readability is very important for productivity, so I'm concerned this pattern is going to move us in the wrong direction by adding friction to the dev workflow. If there are any objective reasons that this pattern is better I think we can look at implementing it for new components. Similarly, if @padala and @domw30 think it is sufficiently readable, we can look into it.
Sorry for responding late here. I agree that code should be readable and debugging should be easier. With too much separation it complicates the development life cycle instead of simplifying it.

ericlowturing commented:
@colbr I just rebased from `develop` and I can see someone updated `MintProvider` (with "Pets" text) which is deprecated with redux and useMint hook. Not sure how it is passed feature review because I can not see `MintProvider` in `Zns.tsx`. So I just updated the text in `useMint.tsx` hook. And I recommend devs to rebase from `develop` branch when start new task. Also I recommend to get rid of warnings (e.g, unused vars) in console. If we don't care it, its hard to maintain and fix bugs later. Thanks

padala commented:
> @colbr I just rebased from `develop` and I can see someone updated `MintProvider` (with "Pets" text) which is deprecated with redux and useMint hook. Not sure how it is passed feature review because I can not see `MintProvider` in `Zns.tsx`. So I just updated the text in `useMint.tsx` hook. And I recommend devs to rebase from `develop` branch when start new task. Also I recommend to get rid of warnings (e.g, unused vars) in console. If we don't care it, its hard to maintain and fix bugs later. Thanks

I think it might be a good idea to delete them or at least mark something as deprecated programmatically so that others are informed about it.

ericlowturing commented:
> > Code readability is very important for productivity, so I'm concerned this pattern is going to move us in the wrong direction by adding friction to the dev workflow. If there are any objective reasons that this pattern is better I think we can look at implementing it for new components. Similarly, if @padala and @domw30 think it is sufficiently readable, we can look into it.
> 
> Sorry for responding late here. I agree that code should be readable and debugging should be easier. With too much separation it complicates the development life cycle instead of simplifying it.

I am thinking about `readability` vs `test coverage` now. Why I split into `3 hooks and utiltiies`? Because we can not test `useState, useEffect, useCallback` hooks from `functional components`. But if we use only `custom hooks (+ utility functions) + JSX` we can test all of  `components + hooks + utilities`.  

And for just those 3 hooks are like this:

1. `useData` => All of `useState` data 
2. 'useHandlers' => All of `useCallback` functions (And I can not see we are using `useCallback` in the most of other components. Not sure why we don't use it. Typically useCallback is helpful when passing callback props to highly optimised child components.)
3. `useLifecycle` => All of `useEffect` hooks
4. `utilities or helpers` => It will help us to test as much as we can, and it will increase the test coverage
....
And we can add more custom hooks when the component is very very complex and need to split the code for readability.

I know it will be difficult to read when someone see this custom hooks at first. But it will be clear if see above 4 comments.

ericlowturing commented:
> > @colbr I just rebased from `develop` and I can see someone updated `MintProvider` (with "Pets" text) which is deprecated with redux and useMint hook. Not sure how it is passed feature review because I can not see `MintProvider` in `Zns.tsx`. So I just updated the text in `useMint.tsx` hook. And I recommend devs to rebase from `develop` branch when start new task. Also I recommend to get rid of warnings (e.g, unused vars) in console. If we don't care it, its hard to maintain and fix bugs later. Thanks
> 
> I think it might be a good idea to delete them or at least mark something as deprecated programmatically so that others are informed about it.

So I deleted all unused variables. It effects the code quality.

domw30 commented:
Hey, just caught up with the above thread. In terms of the hooks, I'd say to maintain a smaller amount of layers for readability, as Subbu said above it will ensure simpler debugging with it being more readable and easier to navigate - and as someone relatively early into my developer career having readable code really makes a difference, for others with similar experience to me or new developers who are onboarding themselves to the codebase. 

I agree with the below statement from Brett. If logic is going to be re-used this tends to be how I determine whether code needs to be extracted to avoid duplication and more noise in the codebase, the same for util functions and components. 
```I think we should use hooks when (1) a component is getting complex, and (2) when some logic needs to be re-used somewhere else```

In terms of the CSS/SCSS, I really enjoy using Styled Components and it's structure, so I lean towards using CSS modules as the pattern is fairly similar to that. Agree with the above, we should move away from in line styling, I like to see as little noise as possible in the return of a function, again one of the reasons why I really enjoy using Styled Components as there is no need for any classNames or hard coded strings so its very clean. Using CSS modules is close to this which I prefer - again for readability reasons AND being able to easily navigate to into the module itself. Agree with Eric that we should use global variables and import.

With regards to console warnings I'd go down the route of removing any unused vars, and restraining from commenting out to avoid the warnings.
